src/main/scala/mesosphere/marathon/api/akkahttp/PathMatchers.scala

```diff
@@ -0,0 +1,95 @@
+package mesosphere.marathon
+package api.akkahttp
+
+import akka.http.scaladsl.model.Uri.Path
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.state.{ Group, PathId, RootGroup, Timestamp }
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.PathMatcher.{ Matched, Unmatched }
+import akka.http.scaladsl.server.PathMatcher1
+
+import scala.annotation.tailrec
+
+object PathMatchers {
+  import akka.http.scaladsl.server.PathMatcher._
+
+  /**
+    * Matches the remaining path and transforms it into task id
+    */
+  final val RemainingTaskId = Remaining.map(s => Task.Id(s))
+
+  /**
+    * Tries to match the remaining path as Timestamp
+    */
+  final val Version = Segment.flatMap(string =>
+    try Some(Timestamp(string))
+    catch { case _: IllegalArgumentException => None }
+  )
+
+  private val marathonApiKeywords = Set("restart", "tasks", "versions")
+
+  /**
+    * Matches everything what's coming before api keywords as PathId
+    */
+  case object AppPathIdLike extends PathMatcher1[PathId] {
+
+    @tailrec final def iter(reversePieces: List[String], remaining: Path, consumedSlash: Option[Path] = None): Matching[Tuple1[PathId]] = remaining match {
+      case slash @ Path.Slash(rest) =>
+        iter(reversePieces, rest, Some(slash))
+      case Path.Segment(segment, rest) if !marathonApiKeywords(segment) =>
+        iter(segment :: reversePieces, rest)
+      case _ if reversePieces.isEmpty =>
+        Unmatched
+      case remaining =>
+        Matched(
+          consumedSlash.getOrElse(remaining),
+          Tuple1(PathId.sanitized(reversePieces.reverse)))
+    }
+
+    override def apply(path: Path) = iter(Nil, path)
+  }
+
+  /**
+    * Given the current root group, only match and consume an existing appId
+    *
+    * This is useful because our v2 API has an unfortunate design decision which leads to ambiguity in our URLs, such as:
+    *
+    *   POST /v2/apps/my-group/restart/restart
+    *
+    * The intention here is to restart the app named "my-group/restart"
+    *
+    * Given the url above, this matcher will only consume "my-group/restart" from the path,
+    * leaving the rest of the matcher to match the rest
+    */
+  case class ExistingAppPathId(rootGroup: () => RootGroup) extends PathMatcher1[PathId] {
+    import akka.http.scaladsl.server.PathMatcher._
+
+    @tailrec final def iter(collected: Vector[String], remaining: Path, group: Group): Matching[Tuple1[PathId]] = remaining match {
+      case Path.Slash(rest) =>
+        iter(collected, rest, group)
+      case Path.Segment(segment, rest) =>
+        val rawPathId = collected :+ segment
+        val pathId = PathId.sanitized(rawPathId, true)
+        group.app(pathId) match {
```

Comment:
Thanks a ton. This is easier to follow. For me at least :). I wish we could just match `"apps" / Segments / "restart"`. Anyhow this seems fine to me.